frontend/src/metabase/actions/containers/ActionCreatorModal/ActionCreatorModal.unit.spec.tsx

```diff
@@ -102,10 +115,210 @@ describe("actions > containers > ActionCreatorModal", () => {
+          fields: {},
+        },
+      });
+
+      userEvent.click(screen.getByRole("button", { name: "Save" }));
+      userEvent.click(screen.getByRole("button", { name: "Create" }));
+
+      await waitFor(() => {
+        expect(history.getCurrentLocation().pathname).toBe(initialRoute);
+      });
+
+      expect(
+        screen.queryByText("Changes were not saved"),
+      ).not.toBeInTheDocument();
+      expect(
+        screen.queryByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).not.toBeInTheDocument();
+    });
+  });
+
+  describe("editing existing action", () => {
+    it("does not show custom warning modal when leaving with no changes via SPA navigation", async () => {
+      const action = ACTION;
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/${action.id}`;
+      const { history } = await setup({ initialRoute, action });
+
+      history.push(actionRoute);
+      await waitForElementToBeRemoved(() =>
+        screen.queryAllByTestId("loading-spinner"),
+      );
+
+      history.goBack();
+
+      expect(
+        screen.queryByText("Changes were not saved"),
+      ).not.toBeInTheDocument();
+      expect(
+        screen.queryByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).not.toBeInTheDocument();
+    });
+
+    it("shows custom warning modal when leaving with unsaved changes via SPA navigation", async () => {
+      const action = ACTION;
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/${action.id}`;
+      const { history } = await setup({ initialRoute, action });
+
+      history.push(actionRoute);
+      await waitForElementToBeRemoved(() =>
+        screen.queryAllByTestId("loading-spinner"),
+      );
+
+      userEvent.type(screen.getByDisplayValue(action.name), "a change");
+      userEvent.tab(); // need to click away from the input to trigger the isDirty flag
+
+      history.goBack();
+
+      expect(screen.getByText("Changes were not saved")).toBeInTheDocument();
+      expect(
+        screen.getByText(
+          "Navigating away from here will cause you to lose any changes you have made.",
+        ),
+      ).toBeInTheDocument();
+    });
+
+    it("does not show custom warning modal when saving changes", async () => {
+      const action = ACTION;
+
+      const initialRoute = `/model/${MODEL.id}/detail/actions`;
+      const actionRoute = `/model/${MODEL.id}/detail/actions/${action.id}`;
+      const { history } = await setup({ initialRoute, action });
+
+      history.push(actionRoute);
+      await waitForElementToBeRemoved(() =>
+        screen.queryAllByTestId("loading-spinner"),
+      );
+
+      await waitFor(() => {
+        expect(screen.getByDisplayValue(action.name)).toBeInTheDocument();
+      });
```

Comment:
Interestingly, this test will flake without this, but not every time, and only if all tests in the file are run (i.e. it not does not flake with `.only`).